### PR TITLE
Add find_path, a function that allows finding a path between two states in a transition system

### PR DESCRIPTION
**Pull request by Mati-AAU**
Closes https://github.com/Ecdar-SW5/Reveaal/issues/41
Checks reachability from a chosen start state to en chosen end state.
Will return the path if the end state is reachable.
Is possible to get the path over gRPC
The path will be a vec<vec<vec<EdgeID>> where the first vec is the specific component, the inner vec are Transitions and the inner inner vec are edge ids in the transitions.
Example:
{Component1{Transition1{Edge1, Edge2}, {Transition2{Egde3, Edge4}}, Component2{Transition3{Edge5, Edge3}, Transition4{Edge7, Edge8}}}

### CODE REVIEW

File: src/System/reachability.rs [hunk 1]
```diff
@@ -0,0 +1,156 @@
+use edbm::zones::OwnedFederation;
+
+use crate::ModelObjects::component::{State, Transition};
+use crate::TransitionSystems::{TransitionSystem, LocationID};
+use std::collections::HashMap;
+
+pub struct Path {
```
Comment (Puvikaran2001) [inline src/System/reachability.rs:7]: Comment the struct

File: src/System/reachability.rs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    //start_state: State,
+    //transition: Transition,
+}
+
+pub fn preliminary_check(
```
Comment (Puvikaran2001) [inline src/System/reachability.rs:12]: Public function: missing documentation

File: src/System/reachability.rs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    start_state: &State,
+    end_state: &State,
+    system: &dyn TransitionSystem
+) -> Result<bool, Box<dyn std::error::Error>> {
+    if !system.get_all_locations().contains(start_state.get_location())
```
Comment (Puvikaran2001) [inline src/System/reachability.rs:17]: Lave en variable som holder på all location I systemet
fx.
let all_locations = system.get_all_locations();

File: src/System/reachability.rs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        {return Err("The transition system does not contain the start location".into())}
+    if !system.get_all_locations().contains(end_state.get_location())
+        {return Err("The transition system does not contain the end location".into())}
+    if let Some(invariants) = end_state.get_location().get_invariants() {
+        if !&end_state.zone_ref().has_intersection(invariants) {
+            return Err("The desired end state is not allowed due to the invariant on this location".into())
+        }
+    }
+
```
Comment (Puvikaran2001) [inline src/System/reachability.rs:26]: Remove whitespace

File: src/System/reachability.rs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+///
+/// If it is reachable, it returns a path.
+///
+/// If it is not reachable, it returns None.
+///
+/// The start state can be omitted with None to use the start state of the transition system.
+///
+///## Checking if a state can reach another:
+///
+/// let is_reachable: bool = match find_path(Some(start_state), end_state, transition_system) {
```
Comment (Puvikaran2001) [inline src/System/reachability.rs:42]: Function returns a Result, that contains a options if ok and a string if error
let is_reachable: bool = match find_path(Some(start_state), end_state, transition_system) {
        Ok(option) => match option{
            Some(path) => true,
            None => false,
        },
        Err(string) => panic!(string),
    };

File: src/System/reachability.rs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        let next_state = frontier_states.pop();
+        // All has been explored if no next state exist
+        if next_state.is_none() {
+            break;
+        }
+        let next_state = next_state.unwrap();
+        if reached_end_state(&next_state, &end_state) {
+            return Ok(Some(Path{}))/* TODO: Return the actual path */
+        }
+        for action in system.get_actions(){
```
Comment (Puvikaran2001) [inline src/System/reachability.rs:110]: Make instead a local variable outside the loop of actions. since every time  get_actions() is called it combine the input actions and output actions. Also the actions would not be changed, so therefore a local variable would be more efficient

**Review by Puvikaran2001 — CHANGES_REQUESTED** [review]

File: src/System/reachability.rs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+use crate::ModelObjects::component::{State, Transition};
+use crate::TransitionSystems::{TransitionSystem, LocationID};
+use std::collections::HashMap;
+
+pub struct Path {
+    //start_state: State,
+    //transition: Transition,
+}
+
+pub fn validate_input(
```
Comment (Mikkel-Boje) [inline src/System/reachability.rs:12]: Public documentation

File: src/System/reachability.rs [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+pub fn preliminary_check(
+    start_state: &State,
+    end_state: &State,
+    system: &dyn TransitionSystem
+) -> Result<bool, Box<dyn std::error::Error>> {
+    if !system.get_all_locations().contains(start_state.get_location())
+        {return Err("The transition system does not contain the start location".into())}
+    if !system.get_all_locations().contains(end_state.get_location())
+        {return Err("The transition system does not contain the end location".into())}
```
[the remark this answers is not in the document]
Comment (Mikkel-Boje) [reply src/System/reachability.rs:20]: Siden du har merget vores kan du også se extract_state hvor jeg tror vi har en invalid_location error som du kan bruge i stedet for den Err du returner nu. Eller bare la' være 😄

File: src/System/reachability.rs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+) -> Result<(), Box<dyn std::error::Error>> {
+    if !system.get_all_locations().contains(start_state.get_location())
+        {return Err("The transition system does not contain the start location".into())}
+    if !system.get_all_locations().contains(end_state.get_location())
+        {return Err("The transition system does not contain the end location".into())}
+
+    Ok(())
+}
+
+fn fails_early(
```
Comment (Mikkel-Boje) [inline src/System/reachability.rs:25]: Tror navnet godt kunne være bedre.

File: src/System/reachability.rs [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+///
+///## Omitting start state:
+///
+/// let is_reachable: bool = match find_path(None, end_state, transition_system) {
+///     Some(path) => true,
+///     None => false
+/// };
+///
+// This is the main function for the reachability query.
+pub fn find_path(
```
Comment (Mikkel-Boje) [inline src/System/reachability.rs:62]: Public documentation

File: src/System/reachability.rs [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    if fails_early(&start_state, &end_state, system){
+        return Ok(None);
+    }
+
+    search_algorithm(&start_state, &end_state, system)
+}
+
+pub fn search_algorithm(
```
Comment (Mikkel-Boje) [inline src/System/reachability.rs:87]: Public Documentation

**Review by Mikkel-Boje — CHANGES_REQUESTED** [review]
Fix

File: src/System/reachability.rs (hunk 7 above)
Comment (Puvikaran2001) [inline src/System/reachability.rs:12]: Should be private

**Review by Puvikaran2001 — CHANGES_REQUESTED** [review]

File: src/System/reachability.rs (hunk 4 above)
Comment (Mati-AAU) [reply src/System/reachability.rs]: this line no longer exists

File: src/System/reachability.rs [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    if let Some(s) = begin_state {
+        start_state = s;
+    } else if let Some(s) = system.get_initial_state() {
+        start_state = s;
+    } else {
+        panic!("No state to start with");
+    }
+
+    match preliminary_check(&start_state, &end_state, system) {
+        Err(msg) => panic!("{}", msg),
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply src/System/reachability.rs]: Code no longer exists

File: src/System/reachability.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply src/System/reachability.rs]: Fixed, no longer public

File: src/System/reachability.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply src/System/reachability.rs:20]: saving in a local variable now

Not sure if we should use that error

File: src/System/reachability.rs [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+    frontier_states.push(start_state.clone());
+    loop {
+        let next_state = frontier_states.pop();
+        // All has been explored if no next state exist
+        if next_state.is_none() {
+            break;
+        }
+        let next_state = next_state.unwrap();
+        if reached_end_state(&next_state, &end_state) {
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply src/System/reachability.rs:107]: fixed

File: src/System/reachability.rs [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+fn take_transition(
+    next_state:  &State,
+    transition: &Transition,
+    frontier_states: &mut Vec<State>,
+    visited_states: &mut HashMap<LocationID, Vec<OwnedFederation>>,
+    system: &dyn TransitionSystem) {
+    let mut new_state = next_state.clone();
+    if transition.use_transition(&mut new_state){
+        new_state.extrapolate_max_bounds(system); // Do we need to do this? consistency check does this
+        let existing_zones: &mut Vec<OwnedFederation> = visited_states.entry(new_state.get_location().id.clone()).or_insert(Vec::new());
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply src/System/reachability.rs:138]: Not possible, entry returns a reference to the element in the hashmap so it must be a reference.

File: src/System/reachability.rs [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+            visited_states.get_mut(&new_state.get_location().id).unwrap().push(new_state.zone_ref().clone());
+            frontier_states.push(new_state);
+        }
+    }
+}
+
+/// Checks if this zone is redundant by being a subset of any other zone
+fn zone_subset_of_existing_zones(
+    new_state: &OwnedFederation,
+    existing_states: & Vec<OwnedFederation>
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply src/System/reachability.rs:150]: fixed

File: src/System/reachability.rs (hunk 6 above)
Comment (Mati-AAU) [reply src/System/reachability.rs:111]: Fixed, it is now saved as a local variable outside the loop

File: src/System/reachability.rs (hunk 7 above)
Comment (Mati-AAU) [reply src/System/reachability.rs]: fixed, chose to make it private instead

File: src/System/reachability.rs (hunk 10 above)
Comment (Mati-AAU) [reply src/System/reachability.rs:63]: it already has documentation, removed the weird comment though

File: src/System/reachability.rs (hunk 11 above)
Comment (Mati-AAU) [reply src/System/reachability.rs:88]: Fixed
Made it private, so you cant call it without first going through the checks from find_path

File: src/System/reachability.rs (hunk 7 above)
Comment (Mati-AAU) [reply src/System/reachability.rs]: fixed

File: src/System/reachability.rs (hunk 9 above)
Comment (Mati-AAU) [reply src/System/reachability.rs:26]: is now called is_trivially_uncreachable

File: src/System/reachability.rs [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+/// If it is not reachable, it returns None.
+///
+/// The start state can be omitted with None to use the start state of the transition system.
+///
+///## Checking if a state can reach another:
+///
+/// let is_reachable: bool = match find_path(Some(start_state), end_state, transition_system) {
+///     Some(path) => true,
+///     None => false
+/// };
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply src/System/reachability.rs:53]: fixed

File: src/System/reachability.rs [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+///
+///## Omitting start state:
+///
+/// let is_reachable: bool = match find_path(None, end_state, transition_system) {
+///     Some(path) => true,
+///     None => false
+/// };
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply src/System/reachability.rs:60]: fixed

File: src/System/reachability.rs [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+    end_state: State,
+    system: &dyn TransitionSystem,
+) -> Result<Option<Path>, String> {
+    let start_state: State;
+    if let Some(s) = begin_state {
+        start_state = s;
+    } else if let Some(s) = system.get_initial_state() {
+        start_state = s;
+    } else {
+        panic!("No state to start with");
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply src/System/reachability.rs:74]: Fixed

File: src/System/reachability.rs [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    if !system.get_all_locations().contains(end_state.get_location())
+        {return Err("The transition system does not contain the end location".into())}
+
+    Ok(())
+}
+
+fn fails_early(
+    start_state: &State,
+    end_state: &State,
+    system: &dyn TransitionSystem
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply src/System/reachability.rs:29]: they are supposed to be used later, set underscore for now to avoid complaint

File: src/System/reachability.rs [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+        // All has been explored if no next state exist
+        if next_state.is_none() {
+            break;
+        }
+        let next_state = next_state.unwrap();
+        if reached_end_state(&next_state, end_state) {
+            return Ok(Some(Path{}))/* TODO: Return the actual path */
+        }
+        for action in &actions {
+            for transition in &system.next_transitions(&next_state.decorated_locations, &action){
```
Comment (Puvikaran2001) [inline src/System/reachability.rs:121]: Clippy suggestion: remove & at action

File: src/System/reachability.rs [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+/// If it is reachable, it returns a path.
+///
+/// If it is not reachable, it returns None.
+///
+/// The start state can be omitted with None to use the start state of the transition system.
+///
+///## Checking if a state can reach another:
+///
+/// ```
+/// let is_reachable: bool = match find_path(Some(start_state), end_state, transition_system) {
```
Comment (Puvikaran2001) [inline src/System/reachability.rs:51]: This code example is behave like as code, and therefore it fails in Test Suite

File: src/System/reachability.rs [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+///        Some(path) => true, 
+///        None => false, 
+///    }, 
+///    Err(string) => panic!(string), 
+///};
+/// ```
+///
+///## Omitting start state:
+/// ```
+/// let is_reachable: bool = match find_path(None, end_state, transition_system) {
```
Comment (Puvikaran2001) [inline src/System/reachability.rs:62]: This code example is behave like as code, and therefore it fails in Test Suite

**Review by Puvikaran2001 — CHANGES_REQUESTED** [review]

File: src/System/reachability.rs (hunk 21 above)
Comment (Mati-AAU) [reply src/System/reachability.rs:50]: Made them ignored. Dont know how to make enough mock data to make it run.

File: src/System/reachability.rs (hunk 22 above)
Comment (Mati-AAU) [reply src/System/reachability.rs:61]: Made them ignored. Dont know how to make enough mock data to make it run.

File: src/System/reachability.rs [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+    let actions = system.get_actions();
+
+    frontier_states.push(start_clone);
+    loop {
+        let next_state = frontier_states.pop();
+        // All has been explored if no next state exist
+        if next_state.is_none() {
+            break;
+        }
+        let next_state = next_state.unwrap();
```
Comment (Puvikaran2001) [inline src/System/reachability.rs:122]: Can be simplified

**Review by Puvikaran2001 — CHANGES_REQUESTED** [review]

File: src/System/reachability.rs [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+struct SubPath {
+    previous_sub_path: Option<Rc<SubPath>>,
+    destination_state: State,
+    transition: Option<Transition>,
+}
+
+fn is_trivially_uncreachable(
+    _start_state: &State,
```
Comment (henneboy) [inline src/System/reachability.rs:21]: Is this parameter used?

Comment (S3JER) [reply src/System/reachability.rs]: fixed

File: .gitignore [hunk 25]
```diff
@@ -2,9 +2,11 @@
 target/
 .idea/*
 .vscode/
+.github
```
Comment (mads256h) [inline .gitignore:5]: Remove this line. We should not ignore .github

File: src/DataReader/mod.rs [hunk 26]
```diff
@@ -3,6 +3,7 @@ pub mod json_reader;
 pub mod json_writer;
 pub mod parse_edge;
 pub mod parse_invariant;
+#[allow(deprecated)]
```
Comment (mads256h) [inline src/DataReader/mod.rs:6]: pls no

File: src/System/reachability.rs [hunk 27]
```diff
@@ -0,0 +1,199 @@
...
+fn search_algorithm(start_state: &State, end_state: &State, system: &dyn TransitionSystem) -> Path {
+    // Apply the invariant of the start state to the start state
+    let mut start_clone = start_state.clone();
+    let start_zone = start_clone.take_zone();
+    let zone = start_clone.decorated_locations.apply_invariants(start_zone);
+    start_clone.set_zone(zone);
+
+    // hashmap linking every location to all its current zones
+    let mut visited_states: HashMap<LocationID, Vec<OwnedFederation>> = HashMap::new();
+    //let mut made_transitions: Vec<SubPath> = Vec::new();
```
Comment (mads256h) [inline src/System/reachability.rs:83]: No commented out code pls

File: src/System/reachability.rs [hunk 28]
```diff
@@ -0,0 +1,199 @@
...
+
+    // List of states that are to be visited
+    let mut frontier_states: Vec<Rc<SubPath>> = Vec::new();
+
+    let mut actions: Vec<String> = system.get_actions().into_iter().collect();
+    actions.sort();
```
Comment (mads256h) [inline src/System/reachability.rs:89]: Do we need to sort the actions?

File: src/System/reachability.rs [hunk 29]
```diff
@@ -0,0 +1,199 @@
...
+            frontier_states.push(Rc::new(SubPath {
+                previous_sub_path: Some(Rc::clone(sub_path)),
+                destination_state: new_state,
+                transition: Some(transition.clone()),
+            }));
+        }
+    }
+}
+
+/// Checks if this zone is redundant by being a subset of any other zone
```
Comment (mads256h) [inline src/System/reachability.rs:163]: Sometimes you use `///`

File: src/System/reachability.rs [hunk 30]
```diff
@@ -0,0 +1,199 @@
...
+) -> bool {
+    for existing_state in existing_states {
+        if new_state.subset_eq(existing_state) {
+            return true;
+        }
+    }
+    false
+}
+
+// Removes everything in existing_zones that is a subset of zone
```
Comment (mads256h) [inline src/System/reachability.rs:176]: And other times you use `//`

File: src/System/reachability.rs [hunk 31]
```diff
@@ -0,0 +1,199 @@
...
+fn remove_existing_subsets_of_zone(
+    new_zone: &OwnedFederation,
+    existing_zones: &mut Vec<OwnedFederation>,
+) {
+    existing_zones.retain(|existing_zone| !existing_zone.subset_eq(new_zone));
+}
+
+fn make_path(sub_path: Rc<SubPath>) -> Path {
```
Comment (mads256h) [inline src/System/reachability.rs:184]: It is also inconsistent what functions have comments and what doesn't

File: src/TransitionSystems/transition_id.rs [hunk 32]
```diff
@@ -12,6 +12,109 @@ pub enum TransitionID {
...
+    ) -> (&TransitionID, usize) {
+        match self {
+            TransitionID::Conjunction(l, r) | TransitionID::Composition(l, r) => {
+                let a = l.get_leaves_helper(current_leaves, index);
+                let b = r.get_leaves_helper(current_leaves, a.1 + 1);
+                (self, b.1)
+            }
+            TransitionID::Quotient(_l, _r) => {
+                let mut curIndex = index;
+                for t in _l {
```
Comment (mads256h) [inline src/TransitionSystems/transition_id.rs:46]: `_l` should not have an underscore as it means it is unused

File: src/TransitionSystems/transition_id.rs [hunk 33]
```diff
@@ -12,6 +12,109 @@ pub enum TransitionID {
...
+                    (_, curIndex) = t.get_leaves_helper(current_leaves, index);
+                }
+                let mut lastIndex = curIndex;
+                for s in _r {
```
Comment (mads256h) [inline src/TransitionSystems/transition_id.rs:50]: Same

File: src/tests/EdgeID/transition_id_tests.rs [hunk 34]
```diff
@@ -0,0 +1,105 @@
...
+        transition_ids: Vec<TransitionID>,
+    ) {
+        let mock_model = Box::new(machineExpression);
+        let mut expected_ids: HashSet<&TransitionID> = HashSet::from_iter(transition_ids.iter());
+        let (_, system) =
+            reachability_test_helper_functions::create_system_recipe_and_machine(*mock_model, path);
+        for loc in system.get_all_locations() {
+            for ac in system.get_actions() {
+                for tran in system.next_transitions(&loc, &ac) {
+                    println!("{}", &tran.id);
```
Comment (mads256h) [inline src/tests/EdgeID/transition_id_tests.rs:87]: Please do not print anything in tests

File: src/tests/reachability/get_leaves_tests.rs [hunk 35]
```diff
@@ -0,0 +1,90 @@
...
+                    vec![TransitionID::Simple("e".to_string()), TransitionID::Simple("f".to_string())]
+                )
+            ),
+            Box::new(TransitionID::Simple("g".to_string()))
+        ),
+        vec![
+            vec!(TransitionID::Simple("a".to_string()), TransitionID::Simple("c".to_string())), 
+            vec!(TransitionID::Simple("b".to_string()), TransitionID::Simple("d".to_string())),
+            vec!(TransitionID::Simple("e".to_string()), TransitionID::Simple("f".to_string())),
+            vec!(TransitionID::Simple("g".to_string()))];
```
Comment (mads256h) [inline src/tests/reachability/get_leaves_tests.rs:85]: Cant these just parse a query instead of doing this manually? It is pretty unwieldy

File: src/tests/reachability/search_algorithm_test.rs [hunk 36]
```diff
@@ -0,0 +1,90 @@
...
+    #[test_case(PATH2, "reachability: Component3 -> [L7](x>4); [L8]()", false; "It is not possible to travel from L7 to L8 due to specified guards")]
+    #[test_case(PATH2, "reachability: Component5 -> [L11](); [L12]()", true; "It is possible to travel from L11 to L12 due to update")]
+    #[test_case(PATH2, "reachability: Component6 -> [L13](); [L15]()", true; "It is possible to travel from L13 to L15 due to the updates at L14")]
+    #[test_case(PATH2, "reachability: Component7 -> [L16](); [L19]()", true; "Overwrite state of location once to reach end state")]
+    #[test_case(PATH2, "reachability: Component8 -> [L20](); [L22]()", true; "Reset clock to reach end state")]
+    #[test_case(PATH2, "reachability: Component7 -> [L16](); [L19](y<2)", false; "Unreachable due to second clock")]
+    #[test_case(PATH2, "reachability: Component3 && Component3 -> [L6, L6](); [L7, L7]()", true; "Simple conjunction")]
+    fn search_algorithm_returns_result(path: &str, query: &str, expected: bool) {
+        //TEMPORARY_MISSING_DECLERATIONS_HACK(path);
```
Comment (mads256h) [inline src/tests/reachability/search_algorithm_test.rs:49]: commented out code

File: src/tests/reachability/search_algorithm_test.rs [hunk 37]
```diff
@@ -0,0 +1,90 @@
...
+    #[test_case(PATH2, "reachability: Component6 -> [L13](); [L15]()", Vec::from(["E12", "E11", "E9", "E10", "E13"]); "Path in Component6 from L13 to L15")]
+    #[test_case(PATH2, "reachability: Component7 -> [L16](); [L19]()", Vec::from(["E11", "E12", "E10"]); "Path in Component7 from L16 to L19")]
+    #[test_case(PATH2, "reachability: Component8 -> [L20](); [L22]()", Vec::from(["E13", "E15", "E14"]); "Path in Component8 from L20 to L22")]
+    #[test_case(PATH2, "reachability: Component9 -> [L23](x>5); [L26]()", Vec::from(["E17", "E18"]); "Path in Component9 from L23 x gt 5 to L26")]
+    #[test_case(PATH2, "reachability: Component9 -> [L23](x<5); [L26]()", Vec::from(["E16", "E19"]); "Path in Component9 from L23 x lt 5 to L26")]
+    #[test_case(PATH2, "reachability: Component3 && Component3 -> [L6, L6](); [L7, L7]()", Vec::from(["E5&&E5"]); "Path in Component3 && Component3 from L6 && L6 to L7 && L7")]
+    fn path_gen_test_correct_path(folder_path: &str, query: &str, expected_path: Vec<&str>) {
+        //TEMPORARY_MISSING_DECLERATIONS_HACK(folder_path);
```
Comment (mads256h) [inline src/tests/reachability/search_algorithm_test.rs:68]: same

**Review by mads256h — CHANGES_REQUESTED** [review]

File: src/DataReader/mod.rs (hunk 26 above)
Comment (S3JER) [reply src/DataReader/mod.rs:6]: yes #HateClippy

File: src/System/reachability.rs (hunk 28 above)
Comment (S3JER) [reply src/System/reachability.rs:89]: Da det er et hashset , så vi er nødt til at have samme orden hver gang så fx den samme query, så mister vi vores test så vi ikke kan finde de mange forskellige kombinationer.

File: src/System/reachability.rs (hunk 29 above)
Comment (S3JER) [reply src/System/reachability.rs:163]: Code dokumentation, så det er muligt at se beskrivelsen af funktionen når man holder musen over den.
Vi bruger // når vi beskriver noget inde i en funktion til at gøre det nemmere at læse.

File: src/System/reachability.rs (hunk 30 above)
Comment (S3JER) [reply src/System/reachability.rs:176]: fixed, skulle have været ///

File: src/System/reachability.rs (hunk 31 above)
Comment (S3JER) [reply src/System/reachability.rs:184]: Vi gør det for det meste ved public og nogle gange for private hvis vi syntes de skal beskrives

File: src/TransitionSystems/transition_id.rs (hunk 32 above)
Comment (S3JER) [reply src/TransitionSystems/transition_id.rs:46]: fixed til l

File: src/TransitionSystems/transition_id.rs (hunk 33 above)
Comment (S3JER) [reply src/TransitionSystems/transition_id.rs:50]: fixed til r

File: src/tests/EdgeID/transition_id_tests.rs (hunk 34 above)
Comment (S3JER) [reply src/tests/EdgeID/transition_id_tests.rs:87]: Den mappe burde ikke engang existere xD, har fjernet den da der er samme test et andet sted som ikke printer noget ud. Dette var en gammel version som var sluppet igennem code review.

Comment (S3JER) [reply src/tests/EdgeID/transition_id_tests.rs:87]: Den mappe burde ikke engang existere xD, har fjernet den da der er samme test et andet sted som ikke printer noget ud. Dette var en gammel version som var sluppet igennem code review.

File: src/tests/reachability/get_leaves_tests.rs (hunk 35 above)
Comment (S3JER) [reply src/tests/reachability/get_leaves_tests.rs:85]: It tests what should be tested

File: src/tests/reachability/search_algorithm_test.rs (hunk 37 above)
Comment (S3JER) [reply src/tests/reachability/search_algorithm_test.rs:68]: fixed

File: src/tests/reachability/search_algorithm_test.rs (hunk 36 above)
Comment (S3JER) [reply src/tests/reachability/search_algorithm_test.rs:49]: fixed

File: src/System/reachability.rs (hunk 30 above)
Comment (S3JER) [reply src/System/reachability.rs]: Should be fixed

File: src/DataReader/mod.rs (hunk 26 above)
Comment (Mikkel-Boje) [reply src/DataReader/mod.rs:6]: It was a suggestion by Sebastian to temporarily fix clippy since no one in this project will be fixing it either way.

File: benches/reachability-bench.rs [hunk 38]
```diff
@@ -0,0 +1,36 @@
...
+    bench_reachability(c, "reachability: Machine -> [L5](y<1); [L5](y<2)");
+    bench_reachability(c, "reachability: Machine -> [L5](); [L5]()");
+    bench_reachability(
+        c,
+        "reachability: Machine || Researcher -> [L5, U0](); [L5, L7]()",
+    );
+    bench_reachability(c, "reachability: Researcher -> [U0](); [L7]()");
+}
+
+criterion_group!(reachability_benches, reachability_benchmarking);
```
Comment (mads256h) [inline benches/reachability-bench.rs:35]: Please set this up with the profiler like the other benchmarks

File: src/ProtobufServer/ecdar_requests/send_query.rs [hunk 39]
```diff
@@ -157,14 +159,58 @@ fn convert_ecdar_result(query_result: &QueryResult) -> Option<ProtobufResult> {
                     reason: "".to_string(),
                     relation: vec![],
                     state: None,
-                    action: "".to_string(), // Empty string is used, when no failing action is available.
+                    action: "".to_string(),
```
Comment (mads256h) [inline src/ProtobufServer/ecdar_requests/send_query.rs:162]: Why did you remove this comment?

File: src/ProtobufServer/ecdar_requests/send_query.rs [hunk 40]
```diff
@@ -157,14 +159,58 @@ fn convert_ecdar_result(query_result: &QueryResult) -> Option<ProtobufResult> {
...
+                &path
+                    .path
+                    .as_ref()
+                    .unwrap()
+                    .iter()
+                    .map(|p| p.id.clone())
+                    .collect(),
+            );
+
+            match protoPath {
```
Comment (mads256h) [inline src/ProtobufServer/ecdar_requests/send_query.rs]: `protoPath` -> `proto_path`

File: src/System/executable_query.rs [hunk 41]
```diff
@@ -57,6 +66,13 @@ fn not_satisfied(query_str: &str) {
     println!("{} -- Property is NOT satisfied", query_str);
 }
 
+fn print_path(path: Vec<Transition>) {
```
Comment (mads256h) [inline src/System/executable_query.rs:69]: You can't just take a reference?
It also is very sus to me that you are printing this.

File: src/System/reachability.rs [hunk 42]
```diff
@@ -0,0 +1,198 @@
...
+
+struct SubPath {
+    previous_sub_path: Option<Rc<SubPath>>,
+    destination_state: State,
+    transition: Option<Transition>,
+}
+
+fn is_trivially_unreachable(end_state: &State, _system: &dyn TransitionSystem) -> bool {
+    if let Some(invariants) = end_state.get_location().get_invariants() {
+        if !&end_state.zone_ref().has_intersection(invariants) {
```
Comment (mads256h) [inline src/System/reachability.rs:22]: Is this & needed? it seems like it is taking the reference to the `bool` from `has_intersection`

File: src/System/reachability.rs [hunk 43]
```diff
@@ -0,0 +1,198 @@
...
+            return true;
```
Comment (mads256h) [inline src/System/reachability.rs:23]: pls no returnerino

File: src/System/reachability.rs [hunk 44]
```diff
@@ -0,0 +1,198 @@
...
+///    Err(string) => panic!(string),
+/// };
+/// ```
+pub fn find_path(
+    start_state: State,
+    end_state: State,
+    system: &dyn TransitionSystem,
+) -> Result<Path, String> {
+    if is_trivially_unreachable(&end_state, system) {
+        return Ok(Path {
```
Comment (mads256h) [inline src/System/reachability.rs:65]: Is the return keyword needed here?

File: src/System/reachability.rs [hunk 45]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+/// Checks if this zone is redundant by being a subset of any other zone
+fn zone_subset_of_existing_zones(
+    new_state: &OwnedFederation,
+    existing_states: &Vec<OwnedFederation>,
+) -> bool {
+    for existing_state in existing_states {
+        if new_state.subset_eq(existing_state) {
+            return true;
```
Comment (mads256h) [inline src/System/reachability.rs:169]: Is the return keyword needed here?

**Review by mads256h — CHANGES_REQUESTED** [review]

File: src/System/reachability.rs (hunk 43 above)
Comment (Mikkel-Boje) [reply src/System/reachability.rs:23]: I would say making an else case here to return implicitly will look worse, and will make it harder to expand this section.

File: src/ProtobufServer/ecdar_requests/send_query.rs (hunk 39 above)
Comment (S3JER) [reply src/ProtobufServer/ecdar_requests/send_query.rs:162]: Misstake has been made xD

File: src/System/reachability.rs (hunk 43 above)
Comment (S3JER) [reply src/System/reachability.rs:23]: Not possible,  gets "Mismatched types expected '()', found bool

File: src/System/reachability.rs (hunk 44 above)
Comment (S3JER) [reply src/System/reachability.rs:65]: yes

File: src/System/reachability.rs (hunk 45 above)
Comment (S3JER) [reply src/System/reachability.rs:169]: yes

File: src/System/executable_query.rs (hunk 41 above)
Comment (S3JER) [reply src/System/executable_query.rs:69]: Is used in main on line 61, where the print_result is used with is a implenentation on the QueryResult enum.
Have fixed so it is a reference

File: src/System/reachability.rs (hunk 42 above)
Comment (S3JER) [reply src/System/reachability.rs:22]: Works without the reference so yes.

**Review by mads256h — APPROVED** [review]

**Review by Archdragon79 — APPROVED** [review]
LGTM

**Review by RedToad0 — APPROVED** [review]
lgmt
